lib/internal/repl/utils.js

```diff
@@ -392,8 +400,242 @@ function setupPreview(repl, contextSymbol, bufferSymbol, active) {
   return { showPreview, clearPreview };
 }
 
+function setupReverseSearch(repl) {
+  // Simple terminals can't use reverse search.
+  if (process.env.TERM === 'dumb') {
+    return { reverseSearch() { return false; } };
+  }
+
+  const alreadyMatched = new Set();
+  const labels = {
+    r: 'bck-i-search: ',
+    s: 'fwd-i-search: '
```

Comment:
nit: Do we need to actually change the labels on key press? I think it might not be needed. ( IMHO, the change while I type seems to be little bit annoying)

Reply:
I am uncertain in what way the labels would change during key presses? It is intentional to indicate the search direction, depending on what the user wants: `<ctrl> + R` for the backwards search and `<ctrl> + S` to search forwards.

Reply:
Yes. I wanted to just keep the label as `search-text` instead. Also changing the label like `failed-fwd-i-search` while I type seems to be bit annoying for the user (visually since the text appears and disappears). Again this is not a blocking comment, just my personal opinion.

Reply:
It provides extra information to the user and that should be very valuable. 

`failed-*-i-search` will only come up in case there's no match and it stays there until the search is either cancelled or the input changes to something that matches. Not indicating that there is no match is IMO not a good idea.

Functionality wise it's almost a 1-1 copy of the ZSH reverse-i-search and ZSH is likely the most intuitive and user friendly shell out there (even apple uses it by default by now).

Reply:
I'm ok to land as it is :) when I played around it was a bit annoying to me, definitely not a blocking comment.